### PR TITLE
Fix tracing of allocations from C

### PR DESCRIPTION
**Pull request by chrisseaton**
We get failures while tracing allocations from C extensions. I think this fixes it, but I'm not keen to try to architect a test case, to be honest.

### CODE REVIEW

File: spec/ruby/library/objectspace/memsize_of_all_spec.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require_relative '../../spec_helper'
+require 'objspace'
+
+describe "ObjectSpace.memsize_of_all" do
```
Comment (eregon) [inline spec/ruby/library/objectspace/memsize_of_all_spec.rb:4]: The docs of this say:
> This method is only expected to work with C Ruby.

But it seems we already implement it.
Anyway I think adding specs for it is fine, it doesn't seem worse than `ObjectSpace.each_object`.

**Comment by eregon** [comment]
In general I feel `ObjectSpace.trace_object_allocations` is crazily expensive, and something we should probably give a performance warning for.
CRuby docs also say:
> Note that this feature introduces a huge performance decrease and huge
memory consumption.

So I think of it as something that should only be used for debugging.
Also notes it invalidates a lot of code when enabling/disabling allocation tracing.

One thing we could do to make it a little bit less horrible would be to store allocation information on the DynamicObject iself, instead of on a separate Hash which leaks all allocations done while enabled.

File: spec/ruby/library/objectspace/memsize_of_all_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  it "returns a non-zero Integer for all objects" do
+    ObjectSpace.memsize_of_all.should be_kind_of(Integer)
+    ObjectSpace.memsize_of_all.zero?.should be_false
```
Comment (eregon) [inline spec/ruby/library/objectspace/memsize_of_all_spec.rb:7]: `ObjectSpace.memsize_of_all.should > 0` seems more expressive and clearer, I'll change to that.

**Review by eregon — APPROVED** [review]

**Comment by chrisseaton** [comment]
`trace_object_allocations_clear` could set a generation flag, and allocation information would be tagged with the current flag.

But as you say, not really worth improving it unless a need is seen. I'm fixing a crash here, that's all.

File: spec/ruby/library/objectspace/memsize_of_all_spec.rb (hunk 1 above)
Comment (chrisseaton) [reply spec/ruby/library/objectspace/memsize_of_all_spec.rb:4]: We implement quite a bit more of `objspace` in random ways.

**Comment by eregon** [comment]
> `trace_object_allocations_clear` could set a generation flag, and allocation information would be tagged with the current flag.

I had the same idea and implemented that.

**Pull request merged**